plugin/audit_log/file_logger.c

```diff
@@ -218,38 +218,45 @@ static char *logname(LOGGER_HANDLE *log, char *buf, size_t buf_len,
 }
 
 
-static int do_rotate(LOGGER_HANDLE *log)
-{
-  char namebuf[FN_REFLEN];
-  int result;
-  unsigned int i;
-  char *buf_old, *buf_new, *tmp;
+static int do_rotate(LOGGER_HANDLE *log) {
+  char new_name_buf[FN_REFLEN];
+  char old_name_buf[FN_REFLEN];
+  char *new_name = NULL;
+  char *old_name = NULL;
+  unsigned int i = 0;
 
-  if (log->rotations == 0)
-    return 0;
+  if (log->rotations == 0) return 0;
 
-  memcpy(namebuf, log->path, log->path_len);
+  memcpy(new_name_buf, log->path, log->path_len);
+  memcpy(old_name_buf, log->path, log->path_len);
 
-  buf_new= logname(log, namebuf, sizeof(namebuf), log->rotations);
-  buf_old= log->path;
-  for (i=log->rotations-1; i>0; i--)
-  {
-    logname(log, buf_old, FN_REFLEN, i);
-    if (!access(buf_old, F_OK) &&
-        (result= my_rename(buf_old, buf_new, MYF(0))))
-      goto exit;
-    tmp= buf_old;
-    buf_old= buf_new;
-    buf_new= tmp;
+  for (i = log->rotations; i > 0; i--) {
+    new_name = logname(log, new_name_buf, sizeof(new_name_buf), i);
+
+    if (i > 1) {
+      old_name = logname(log, old_name_buf, sizeof(new_name_buf), i - 1);
+    }
+    else {
+      old_name = old_name_buf;
+      old_name_buf[log->path_len] = 0;
+    }
+
+    if (0 == access(old_name, F_OK) &&
+        0 != my_rename(old_name, new_name, MYF(0))) {
```

Comment:
Is it OK that we rename currently open file? (when `i==1`)

Reply:
Yes, server will continue writing into this renamed file until it gets closed below.